stock.py

```diff
@@ -23,86 +24,66 @@ def __setup__(cls):
     __name__ = 'stock.lot'
 
-    cost_lines = fields.One2Many('stock.lot.cost_line', 'lot', 'Cost Lines')
-    cost_price = fields.Function(fields.Numeric('Cost Price'),
-        'get_cost_price')
-
-    def get_cost_price(self, name):
-        if not self.cost_lines:
-            return
-        return sum(l.unit_price for l in self.cost_lines if l.unit_price is not
-            None)
-
-    @fields.depends('product', 'cost_lines')
-    def on_change_product(self):
-        try:
-            super(Lot, self).on_change_product()
-        except AttributeError:
-            pass
-
-        if not self.id or self.id <= 0:
-            return
-
-        cost_lines = self._on_change_product_cost_lines()
-        if cost_lines:
-            cost_lines = cost_lines.get('add')
-            LotCostLine = Pool().get('stock.lot.cost_line')
-            lot_cost_lines = LotCostLine.search([
-                    ('lot', '=', self.id),
-                    ('category', '=', cost_lines[0][1]['category']),
-                    ('unit_price', '=', cost_lines[0][1]['unit_price']),
-                    ])
-            if lot_cost_lines:
-                self.cost_lines = lot_cost_lines
-
-    def _on_change_product_cost_lines(self):
+    cost_price = fields.Function(fields.Numeric("Cost Price"),
+        'get_lot_prices')
+    total_cost = fields.Function(fields.Numeric("Total Cost"),
+        'get_lot_prices')
+
+    @classmethod
+    def get_lot_prices(cls, lots, names):
         pool = Pool()
-        ModelData = pool.get('ir.model.data')
-
-        if not self.product:
-            return {}
-
-        category_id = ModelData.get_id('stock_lot_cost',
-            'cost_category_standard_price')
-        return {
-            'add': [(0, {
-                        'category': category_id,
-                        'unit_price': self.product.cost_price,
-                        })],
-            }
+        Move = pool.get('stock.move')
+        Product = pool.get('product.product')
+        Location = pool.get('stock.location')
+
+        res = {}
+        ids = [x.id for x in lots]
+
+        for name in ['cost_price', 'total_cost']:
+            res[name] = dict.fromkeys(ids)
+
+        warehouse_ids = [location.id for location in Location.search(
+            [('type', '=', 'warehouse')])]
+        product_ids = list(set(lot.product.id for lot in lots))
+        lot_ids = list(set(lot.id for lot in lots))
+
+        with Transaction().set_context({'stock_date_end': datetime.date.max}):
+            pbl = Product.products_by_location(warehouse_ids,
+                with_childs=True,
+                grouping=('product', 'lot'),
+                grouping_filter=(product_ids, lot_ids))
+
+        for lot in lots:
+            moves = Move.search([
+                ('lot', '=', lot.id),
+                ('origin', 'not like', 'stock.move,%'),
+                ('from_location.type', '=', 'supplier'),
+                ('to_location.type', '=', 'storage'),
```

Comment:
You must filter by state = 'done'.

Also, you should use search() outside the loop filtering by `lot in [lots]` and then grouping.